### PR TITLE
Add support for int-mask and int-mask-of types

### PR DESCRIPTION
**Pull request by rvanvelzen**
Closes phpstan/phpstan#4732

### CODE REVIEW

**Comment by rvanvelzen** [comment]
I'm not convinced 0 should be part of the union by default, but it's easy enough to add. Psalm does include it, so for parity it might be best to include it.

**Comment by ondrejmirtes** [comment]
Yeah, let's include `0`. It's questionable, in some cases people want `0` and in some cases they don't, but we don't know which is which so let's do the wider and less annoying thing here.

File: src/PhpDoc/TypeNodeResolver.php [hunk 1]
```diff
@@ -833,6 +852,32 @@ private function resolveConstTypeNode(ConstTypeNode $typeNode, NameScope $nameSc
...
+		}
+
+		$values = [];
+
+		foreach ($ints as $int) {
+			$int = $int->getValue();
+
+			if ($int !== 0) {
+				foreach ($values as $value) {
+					$values[] = $value | $int;
```
Comment (ondrejmirtes) [inline src/PhpDoc/TypeNodeResolver.php]: One idea for improvement. If the generated integer list is continuous, we could return IntegerRangeType instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Also please submit documentation for these types to https://phpstan.org/writing-php-code/phpdoc-types - there's a link to edit the page at the bottom.

**Comment by ondrejmirtes** [comment]
Perfect, thank you!

Two ideas:

1) Test for a scenario when a value that's not power-of-2 (like 3) is part of the values.
2) when a non-integer is mentioned in the PHPDoc, the resulting type should be ErrorType. That will get picked up as unresolvable type by a rule.

**Comment by Seldaek** [comment]
Great, thanks!

**Pull request merged**